[DependencyInjection][FrameworkBundle] Fix using PHP 8.1 enum as parameters

ogizanagi opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 4.4 <!-- see below -->
| Bug fix?      | yesish
| New feature?  | no <!-- please update src/**/CHANGELOG.md files -->
| Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Tickets       | Fix #44834  <!-- prefix each issue number with "Fix #", no need to create an issue if none exist, explain below instead -->
| License       | MIT
| Doc PR        | N/A

While #40857 allowed using enums in DI, it does not allow using these in parameters.

This would be the actual fix for #44834, which shows the error you'll get trying to use enum as DI parameters (appart from the binding issue fixed in #44838):

![image](https://user-images.githubusercontent.com/2211145/147762495-f57f1fff-f56a-4e87-bbc4-3bba97a8e81b.png)

given:

```php
namespace App;

enum Status {
    case Draft;
    case Deleted;
    case Published;
}
```

```yaml
# services.yaml
parameters:
    default_status: !php/const App\Status::Draft
```

The exception happens because the `PhpDumper` misses the leading slash:

```diff

    protected function getDefaultParameters(): array
    {
        return [
-            'default_status' => App\Status::Draft,
+            'default_status' => \App\Status::Draft,
        ];
    }
```

While this would fix using enums as DI parameters as of Symfony < 6, 
the `ParameterBagInterface::get/set` and `ContainerInterface::setParameter/getParameter` are not allowing `\UnitEnum` and adding these in phpdoc is an issue since actual typehints and return types were added as of Symfony 6.
=> So this PR is a BC break, but hopefully nobody will be hit by it 🤞🏻 

(poke @ismail1432 - 
https://twitter.com/SmaineDev/status/1476237072826613763?s=20)

derrabus commented:
Either way, allowing enums in parameters would be a new feature. I don't think we should add that functionality to 4.4.

ogizanagi commented:
I kind of agree, despite https://github.com/symfony/symfony/pull/40857 being on 4.4. But the difference is parameters never supported objects.
So let's make that clear by throwing in 4.4 and see what we can do/if we want to support enum in parameters later?

derrabus commented:
Sounds good. 🙂

nicolas-grekas commented:
How do we move forward on this one?

ogizanagi commented:
Sorry, didn't have time to go further. The plan was to throw either in the `ParameterBag` or the `PhpDumper` so that's explicit enum types are not supported as parameters ; since it even goes against the `ParameterBagInterface::get/set` contract.

ogizanagi commented:
Updated the PR so that we simply throw on attempting to set an enum or array containing an enum as a DI parameter : this should be enough for our goal: explicit the fact enums aren't supported as such.
Of course, the phpdoc already tells this, and we won't check for any other infringement to it ; but since the DI loaders allow such a thing indirectly:

```yml
# services.yaml
parameters:
    default_status: !php/const App\Status::Draft
    statuses: 
        - !php/const App\Status::Draft
        - !php/const App\Status::Deleted
        - !php/const App\Status::Published
```

we might want to handle this specific case here so the user notices it.

nicolas-grekas commented:
I'm sorry I'm not super convinced by this. It would be strange to me to allow direct references to enums in service arguments, but to disallow indirect references via DI parameters:

```yaml
parameters:
    foo: !php/const Some\Enum::Case

services:
    bar: ['%foo%']
```

I'd prefer to fix supporting them when dumping, and widen the types on 6.0. I don't think anybody is going to be hit by the BC break.

ogizanagi commented:
Honestly, I'd be way more satisfied this way too :)
I'll revert to the previous state of this PR.

ogizanagi commented:
Alright, PR updated. Hopefully I haven't missed anything.

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Tests/Console/Descriptor/ObjectsProvider.php:
```diff
@@ -61,14 +63,24 @@ public static function getRoutes()
+            'array' => [12, 'Hello world!', true],
+        ]);
+
+        if (\PHP_VERSION_ID >= 80100) {
```
let's return early and unindent the rest to reduce merge conflicts

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Tests/Dumper/PhpDumperTest.php:
```diff
@@ -1217,16 +1218,37 @@ public function testDumpHandlesEnumeration()
+        return [
+            'unit_enum' => \Symfony\Component\DependencyInjection\Tests\Fixtures\FooUnitEnum::BAR,
+            'enum_array' => [
+                0 => \Symfony\Component\DependencyInjection\Tests\Fixtures\FooUnitEnum::BAR,
```
I think these should be part of `getDynamicParameter` instead, so that `getDefaultParameters` keeps returning a static array with no side effects (not even class loading)

ogizanagi replied:
Done in f0acb77663bf3613877e7ad707a788bffadb5b81

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php:
```diff
@@ -1545,12 +1546,12 @@ protected function getDefaultParameters(): array
      * @throws InvalidArgumentException
      */
-    private function exportParameters(array $parameters, string $path = '', int $indent = 12): string
+    private function exportParameters(array $parameters, string $path = '', int $indent = 12, &$hasEnum = false): string
```
`bool` type hint?

ogizanagi replied:
Oops. Added

fabpot commented:
We stopped adding new features for supporting newer PHP versions a while ago, so this is a new feature that should go into 6.1.

nicolas-grekas commented:
This is edgy, but after a private chat with @fabpot, I've got a go for merging. Now doing it.

nicolas-grekas commented:
Thank you @ogizanagi.
